Kernel/Graphics/VirtIOGPU/DisplayConnector.cpp

```diff
@@ -0,0 +1,788 @@
+#include <Kernel/Graphics/VirtIOGPU/GraphicsAdapter.h>
+#include <Kernel/Graphics/VirtIOGPU/Protocol.h>
+#include <Kernel/Random.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+namespace Kernel {
+
+VirtIODisplayConnector::PerContextState::PerContextState(Graphics::VirtIOGPU::ContextID context_id, NonnullOwnPtr<Memory::Region> transfer_buffer_region)
+    : m_context_id(context_id)
+    , m_transfer_buffer_region(move(transfer_buffer_region))
+{
+}
+
+void VirtIODisplayConnector::detach(OpenFileDescription& description)
+{
+    m_context_state_lookup.remove(&description);
+    DisplayConnector::detach(description);
+}
+
+ErrorOr<RefPtr<VirtIODisplayConnector::PerContextState>> VirtIODisplayConnector::get_context_for_description(OpenFileDescription& description)
+{
+    auto res = m_context_state_lookup.get(&description);
+    if (!res.has_value())
+        return EBADF;
+    return res.value();
+}
+
+NonnullRefPtr<VirtIODisplayConnector> VirtIODisplayConnector::must_create(VirtIOGraphicsAdapter& graphics_adapter, Graphics::VirtIOGPU::ScanoutID scanout_id)
+{
+    // Setup memory transfer region
+    auto virtio_virgl3d_uploader_buffer_region = MUST(MM.allocate_kernel_region(
+        NUM_TRANSFER_REGION_PAGES * PAGE_SIZE,
+        "VIRGL3D upload buffer",
+        Memory::Region::Access::ReadWrite,
+        AllocationStrategy::AllocateNow));
+
+    auto scratch_space_region = MUST(MM.allocate_contiguous_kernel_region(
+        32 * PAGE_SIZE,
+        "VirtGPU Scratch Space",
+        Memory::Region::Access::ReadWrite));
+
+    auto device_or_error = DeviceManagement::try_create_device<VirtIODisplayConnector>(graphics_adapter, scanout_id, move(scratch_space_region), move(virtio_virgl3d_uploader_buffer_region));
+    VERIFY(!device_or_error.is_error());
+    auto connector = device_or_error.release_value();
+    connector->initialize_3d_context();
+    MUST(connector->set_safe_resolution());
+    connector->initialize_console();
+    return connector;
+}
+
+void VirtIODisplayConnector::initialize_3d_context()
+{
+    if (!m_graphics_adapter->virgl_feature_accepted())
+        return;
+    m_kernel_context_id = create_context();
+}
+
+VirtIODisplayConnector::VirtIODisplayConnector(VirtIOGraphicsAdapter& graphics_adapter, Graphics::VirtIOGPU::ScanoutID scanout_id, NonnullOwnPtr<Memory::Region> scratch_space_region, NonnullOwnPtr<Memory::Region> virtio_virgl3d_uploader_buffer_region)
+    : DisplayConnector()
+    , m_graphics_adapter(graphics_adapter)
+    , m_scanout_id(scanout_id)
+    , m_scratch_space(move(scratch_space_region))
+    , m_virgl3d_uploader_buffer_region(move(virtio_virgl3d_uploader_buffer_region))
+{
+}
+
+void VirtIODisplayConnector::initialize_console()
+{
+    m_console = Kernel::Graphics::VirtIOGPU::Console::initialize(*this);
+}
+
+DisplayConnector::Hardware3DAccelerationCommandSet VirtIODisplayConnector::hardware_3d_acceleration_commands_set() const
+{
+    return DisplayConnector::Hardware3DAccelerationCommandSet::VirGL;
```

Comment:
Nit: This probably shouldn't return VirGL if VirGL support wasn't negotiated between the Graphics Adapter and the hardware.